Feature/binary tempfiles

antbern opened the pull request:
After hunting performance for some time (including trying to cache the temporary files in memory) without any luck, I decided to give an old idea a go. We had a short talk about it in discussion  #121 and it did not really lead anywhere at the time, but  now I decided to give it a go :racehorse: 

Short summary;
- Replace text-formatted line-oriented XYZ files with custom binary format that simply writes the points one by one into a byte stream. No library needed and performance seems to be alright! This means we no longer need to do much of two things: 1) format `f64` values and write that to a file, and 2) read that file line-by-line and parse the `f64` again. This gives _huge_ performance improvements!
- Added a command `internal2xyz` to convert the new internal binary format back into the previous XYZ line-oriented format. Useful if anyone still want to use the temporary files for other processing :rocket: 
- Disable the `aes` feature of `zip` crate since we do not deal with password protected zip files anyway.
- Added some more files to the `.gitignore` for convenience.

Observed Results:
- Regression jobs execution time down to `~60`  seconds from previous `~94` seconds, almost more than 33% reduction! :fire:
- Size of temporary files reduced from `~1.5GB` to `~1.0GB` (xyztemp.xyz)  :100: 
- No change in memory usage.

I would say it is a great leap forward in terms of large-scale processing feasibility :rocket:

carlogico commented:
@rphlo I was about to post the same question in https://github.com/rphlo/karttapullautin/discussions/121. It looks to me like  x y z coordinates would not suffer from being stored as f32. If the math requires more precision the coordinates could be just cast as f64 on read.

antbern commented:
> @antbern I have been wondering about the choice to use f64 instead of f32 for xyz data and general math in karttapullautin. Would the gain memory wise and storage wise be worth it?

Yes that would basically half the storage, loading, and memory requirements during processing. Not sure how much of a difference it would do on the actual computation time, however. And then there is the possible issue with precision. A `f32` can generally represent 7 digits accurately which means that depending on the coordinate system used for the data you might run into issues or not. Not sure if it is worth taking that risk.

I did do a test awhile back and changed all to `f32` for just this reason and, well it _did change the output_. Some contour lines were different and the vegetation also changed. In the end it is of course a tradeoff, although I'm not sure it's worth it in this case. Depends also on how much precision the las/laz files provide. I read somewhere that they use `i16` to store coordinates internally, with some floating point offsets and scaling, and if this is not up to 7 digits then it might be okay to use it for storage only as @carlogico mention. IMO it is worth keeping the `f64` just to not run into weird to debug edge cases that only affect certain coordinate systems in special situations :smile:

carlogico commented:
I wouldn't expect any gains in computation time.
I'd argue this:
* the altitude data would most likely not suffer from being f32 (even if it was in feet)
* I can see the concern about the x and y coordinates. This could be mitigated by having those always stored starting at 0 (and maybe storing xstart and ystart as the first record in the file), and potentially a multiplicative factor. Having the bounds stored somewhere would also obviate the need to iterate through the entire file just to get those.
* in addition to that, the creation of xyz2 and xyz_03 seems to happen on a grid, which suggests that a real32 png assisted by a corresponding pgw file should be sufficient.

antbern commented:
> * in addition to that, the creation of xyz2 and xyz_03 seems to happen on a grid, which suggests that a real32 png assisted by a corresponding pgw file should be sufficient.

@carlogico this is a good observation. I've started work on storing and reading this information as a grid instead of in the XYZ point-like format in https://github.com/antbern/rusty-pullauta/pull/28. It does clean up a lot (since we also store the offsets and scale in the same fil) and will allow optimizing the computations in the future (we can directly index a `Vec` instead of doing `HashMap` lookups). As for the `f64` vs `f32` discussion, for now the file is not more than 20-40MB so I believe I will stick with the `f64` for now. We can always experiment later once this cleanup is done :rocket: Will open a PR once that is done.

carlogico commented:
Those changes look very promising. I agree that given those, the `f64` vs `f32` becomes much less impactful.
Another thing I noticed (disclaimer:  I haven't fully digested the code...), is that the xyz2 and xyz_02 files are being read many times in the main processing pipeline. Why not just keep the data in memory and pass that to the relevant functions instead of the filename? The data could be saved just at the end of the pipeline anyways. I don't think there should be a memory penalty (those hash maps or `Vec2D`s are created anyways in the relevant functions), and it would save a good amount of i/o.

carlogico commented:
Of course. I’m only talking about the instances where the files are read in memory anyways, so the ram usage would not really change
